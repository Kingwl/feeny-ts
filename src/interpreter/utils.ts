import { assertDef } from '../utils';
import { VarValues, CodeValues } from './types';
import { BaseValue } from './values';

export function assertThisValue<T extends BaseValue>(
  value: T | undefined
): asserts value is NonNullable<T> {
  assertDef(value, "'This' cannot be null");
}

export function assertArgumentsLength(expected: number, actual: number) {
  if (expected !== actual) {
    throw new Error(
      `Arguments mis match, expected ${expected}, actual: ${actual}`
    );
  }
}

export function isVarValues(value: BaseValue): value is VarValues {
  return (
    value.isInteger() ||
    value.isNull() ||
    value.isObject() ||
    value.isArray() ||
    value.isString()
  );
}

export function isCodeValues(value: BaseValue): value is CodeValues {
  return value.isFunction();
}
