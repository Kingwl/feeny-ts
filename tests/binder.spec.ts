import { forEachCases, bindCode } from './utils';

describe('Binder', () => {
  forEachCases((baseName, content) => {
    const { localsResult, membersResult } = bindCode(content);

    it(`${baseName} - locals`, () => {
      expect(localsResult).toMatchSnapshot();
    });

    it(`${baseName} - members`, () => {
      expect(membersResult).toMatchSnapshot();
    });
  });
});
