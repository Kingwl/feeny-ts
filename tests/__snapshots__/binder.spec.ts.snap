// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Binder should-work-with-arrays - locals 1`] = `Array []`;

exports[`Binder should-work-with-arrays - members 1`] = `Array []`;

exports[`Binder should-work-with-binary-shorthand - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 7,
        "name": "x",
      },
      Object {
        "declarationPos": 10,
        "flags": "Variable",
        "id": 8,
        "name": "y",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-binary-shorthand - members 1`] = `Array []`;

exports[`Binder should-work-with-break-and-continue - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-break-and-continue - members 1`] = `Array []`;

exports[`Binder should-work-with-function - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Function",
        "id": 22,
        "name": "f",
      },
    ],
  },
  Object {
    "kind": "FunctionStatement",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 7,
        "flags": "Parameter",
        "id": 26,
        "name": "x",
      },
      Object {
        "declarationPos": 10,
        "flags": "Parameter",
        "id": 27,
        "name": "y",
      },
      Object {
        "declarationPos": 13,
        "flags": "Parameter",
        "id": 28,
        "name": "z",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-function - members 1`] = `Array []`;

exports[`Binder should-work-with-function-call - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Function",
        "id": 10,
        "name": "f",
      },
    ],
  },
  Object {
    "kind": "FunctionStatement",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 7,
        "flags": "Parameter",
        "id": 12,
        "name": "v",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-function-call - members 1`] = `Array []`;

exports[`Binder should-work-with-function-expression - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 16,
        "name": "fn",
      },
    ],
  },
  Object {
    "kind": "FunctionExpression",
    "pos": 9,
    "symbols": Array [
      Object {
        "declarationPos": 18,
        "flags": "Parameter",
        "id": 20,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-function-expression - members 1`] = `Array []`;

exports[`Binder should-work-with-get-shorthand - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-get-shorthand - members 1`] = `Array []`;

exports[`Binder should-work-with-global-variable - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-global-variable - members 1`] = `Array []`;

exports[`Binder should-work-with-if-expression - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-if-expression - members 1`] = `Array []`;

exports[`Binder should-work-with-if-expression-with-if-only - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-if-expression-with-if-only - members 1`] = `Array []`;

exports[`Binder should-work-with-integer - locals 1`] = `Array []`;

exports[`Binder should-work-with-integer - members 1`] = `Array []`;

exports[`Binder should-work-with-local-variable - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Function",
        "id": 16,
        "name": "foo",
      },
    ],
  },
  Object {
    "kind": "SequenceOfStatements",
    "pos": 16,
    "symbols": Array [
      Object {
        "declarationPos": 16,
        "flags": "Variable",
        "id": 20,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-local-variable - members 1`] = `Array []`;

exports[`Binder should-work-with-method-call - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 46,
        "name": "o",
      },
    ],
  },
  Object {
    "kind": "MethodSlot",
    "pos": 20,
    "symbols": Array [
      Object {
        "declarationPos": 29,
        "flags": "Parameter",
        "id": 60,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-method-call - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "MethodSlot",
        "id": 55,
        "name": "f",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-nested-function - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 61,
        "name": "m",
      },
      Object {
        "declarationPos": 10,
        "flags": "Function",
        "id": 62,
        "name": "foo",
      },
    ],
  },
  Object {
    "kind": "FunctionStatement",
    "pos": 10,
    "symbols": Array [
      Object {
        "declarationPos": 19,
        "flags": "Parameter",
        "id": 72,
        "name": "x",
      },
    ],
  },
  Object {
    "kind": "SequenceOfStatements",
    "pos": 27,
    "symbols": Array [
      Object {
        "declarationPos": 27,
        "flags": "Variable",
        "id": 77,
        "name": "y",
      },
      Object {
        "declarationPos": 41,
        "flags": "Function",
        "id": 78,
        "name": "bar",
      },
    ],
  },
  Object {
    "kind": "FunctionStatement",
    "pos": 41,
    "symbols": Array [
      Object {
        "declarationPos": 50,
        "flags": "Parameter",
        "id": 80,
        "name": "z",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-nested-function - members 1`] = `Array []`;

exports[`Binder should-work-with-null - locals 1`] = `Array []`;

exports[`Binder should-work-with-null - members 1`] = `Array []`;

exports[`Binder should-work-with-object-without-extends - locals 1`] = `Array []`;

exports[`Binder should-work-with-object-without-extends - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 12,
        "flags": "VariableSlot",
        "id": 37,
        "name": "x",
      },
      Object {
        "declarationPos": 27,
        "flags": "MethodSlot",
        "id": 38,
        "name": "f",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-objects - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 52,
        "name": "p",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-objects - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "VariableSlot",
        "id": 57,
        "name": "y",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 32,
    "symbols": Array [
      Object {
        "declarationPos": 47,
        "flags": "VariableSlot",
        "id": 65,
        "name": "x",
      },
      Object {
        "declarationPos": 62,
        "flags": "MethodSlot",
        "id": 66,
        "name": "f",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-objects-and-slots - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 124,
        "name": "p",
      },
    ],
  },
  Object {
    "kind": "MethodSlot",
    "pos": 62,
    "symbols": Array [
      Object {
        "declarationPos": 71,
        "flags": "Parameter",
        "id": 162,
        "name": "x",
      },
      Object {
        "declarationPos": 74,
        "flags": "Parameter",
        "id": 163,
        "name": "y",
      },
      Object {
        "declarationPos": 77,
        "flags": "Parameter",
        "id": 164,
        "name": "z",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-objects-and-slots - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "VariableSlot",
        "id": 129,
        "name": "y",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 32,
    "symbols": Array [
      Object {
        "declarationPos": 47,
        "flags": "VariableSlot",
        "id": 149,
        "name": "x",
      },
      Object {
        "declarationPos": 62,
        "flags": "MethodSlot",
        "id": 150,
        "name": "f",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-printf - locals 1`] = `Array []`;

exports[`Binder should-work-with-printf - members 1`] = `Array []`;

exports[`Binder should-work-with-set-shorthand - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-set-shorthand - members 1`] = `Array []`;

exports[`Binder should-work-with-slot-assignment - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 403,
        "name": "o",
      },
      Object {
        "declarationPos": 35,
        "flags": "Variable",
        "id": 410,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-slot-assignment - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "VariableSlot",
        "id": 408,
        "name": "field",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 43,
    "symbols": Array [
      Object {
        "declarationPos": 55,
        "flags": "VariableSlot",
        "id": 475,
        "name": "b",
      },
      Object {
        "declarationPos": 79,
        "flags": "VariableSlot",
        "id": 499,
        "name": "c",
      },
      Object {
        "declarationPos": 107,
        "flags": "VariableSlot",
        "id": 505,
        "name": "d",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 63,
    "symbols": Array [
      Object {
        "declarationPos": 79,
        "flags": "VariableSlot",
        "id": 523,
        "name": "c",
      },
      Object {
        "declarationPos": 107,
        "flags": "VariableSlot",
        "id": 529,
        "name": "d",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 87,
    "symbols": Array [
      Object {
        "declarationPos": 107,
        "flags": "VariableSlot",
        "id": 535,
        "name": "d",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-slot-lookup - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 403,
        "name": "o",
      },
      Object {
        "declarationPos": 35,
        "flags": "Variable",
        "id": 410,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-slot-lookup - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "VariableSlot",
        "id": 408,
        "name": "field",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 43,
    "symbols": Array [
      Object {
        "declarationPos": 55,
        "flags": "VariableSlot",
        "id": 475,
        "name": "b",
      },
      Object {
        "declarationPos": 79,
        "flags": "VariableSlot",
        "id": 499,
        "name": "c",
      },
      Object {
        "declarationPos": 107,
        "flags": "VariableSlot",
        "id": 505,
        "name": "d",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 63,
    "symbols": Array [
      Object {
        "declarationPos": 79,
        "flags": "VariableSlot",
        "id": 523,
        "name": "c",
      },
      Object {
        "declarationPos": 107,
        "flags": "VariableSlot",
        "id": 529,
        "name": "d",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 87,
    "symbols": Array [
      Object {
        "declarationPos": 107,
        "flags": "VariableSlot",
        "id": 535,
        "name": "d",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-arrays - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-arrays - members 1`] = `Array []`;

exports[`Binder should-work-with-type-function-call - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Function",
        "id": 19,
        "name": "add",
      },
      Object {
        "declarationPos": 57,
        "flags": "Variable",
        "id": 24,
        "name": "a",
      },
    ],
  },
  Object {
    "kind": "FunctionStatement",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 9,
        "flags": "Parameter",
        "id": 22,
        "name": "x",
      },
      Object {
        "declarationPos": 21,
        "flags": "Parameter",
        "id": 23,
        "name": "y",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-function-call - members 1`] = `Array []`;

exports[`Binder should-work-with-type-function-expression - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-function-expression - members 1`] = `Array []`;

exports[`Binder should-work-with-type-function-parameter - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Function",
        "id": 10,
        "name": "foo",
      },
    ],
  },
  Object {
    "kind": "FunctionStatement",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 9,
        "flags": "Parameter",
        "id": 12,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-function-parameter - members 1`] = `Array []`;

exports[`Binder should-work-with-type-function-statement - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Function",
        "id": 4,
        "name": "foo",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-function-statement - members 1`] = `Array []`;

exports[`Binder should-work-with-type-integer - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-integer - members 1`] = `Array []`;

exports[`Binder should-work-with-type-method-signature - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "TypeDef",
        "id": 16,
        "name": "foo",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-method-signature - members 1`] = `
Array [
  Object {
    "kind": "TypeDefDeclaration",
    "name": "foo",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 17,
        "flags": "Variable",
        "id": 17,
        "name": "a",
      },
      Object {
        "declarationPos": 36,
        "flags": "MethodSlot",
        "id": 18,
        "name": "f",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-method-slot - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 58,
        "name": "a",
      },
    ],
  },
  Object {
    "kind": "MethodSlot",
    "pos": 45,
    "symbols": Array [
      Object {
        "declarationPos": 56,
        "flags": "Parameter",
        "id": 76,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-method-slot - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "VariableSlot",
        "id": 69,
        "name": "foo",
      },
      Object {
        "declarationPos": 45,
        "flags": "MethodSlot",
        "id": 70,
        "name": "bar",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-null - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-null - members 1`] = `Array []`;

exports[`Binder should-work-with-type-typedef - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "TypeDef",
        "id": 10,
        "name": "foo",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-typedef - members 1`] = `
Array [
  Object {
    "kind": "TypeDefDeclaration",
    "name": "foo",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 17,
        "flags": "Variable",
        "id": 11,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-variable-reference - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "TypeDef",
        "id": 31,
        "name": "foo",
      },
      Object {
        "declarationPos": 33,
        "flags": "Variable",
        "id": 34,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-variable-reference - members 1`] = `
Array [
  Object {
    "kind": "TypeDefDeclaration",
    "name": "foo",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 17,
        "flags": "Variable",
        "id": 32,
        "name": "a",
      },
    ],
  },
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 46,
    "symbols": Array [
      Object {
        "declarationPos": 58,
        "flags": "VariableSlot",
        "id": 39,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-variable-slot - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 22,
        "name": "a",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-type-variable-slot - members 1`] = `
Array [
  Object {
    "kind": "ObjectsExpression",
    "name": undefined,
    "pos": 8,
    "symbols": Array [
      Object {
        "declarationPos": 20,
        "flags": "VariableSlot",
        "id": 27,
        "name": "foo",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-variable-assignment - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-variable-assignment - members 1`] = `Array []`;

exports[`Binder should-work-with-variable-reference - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 4,
        "name": "x",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-variable-reference - members 1`] = `Array []`;

exports[`Binder should-work-with-while-expression - locals 1`] = `
Array [
  Object {
    "kind": "SequenceOfStatements",
    "pos": 0,
    "symbols": Array [
      Object {
        "declarationPos": 0,
        "flags": "Variable",
        "id": 7,
        "name": "x",
      },
      Object {
        "declarationPos": 10,
        "flags": "Variable",
        "id": 8,
        "name": "y",
      },
    ],
  },
]
`;

exports[`Binder should-work-with-while-expression - members 1`] = `Array []`;
